syntax = "proto2";

package broker;

option java_multiple_files = true;
option java_package = "proto.broker";
option java_outer_classname = "BrokerProto";

message EmptyMessage{

}

message EmptyIdRequest{
  required int64 idRequester = 1;
}

message HandshakeRequest{
  required string type = 1;
  required string ip = 2;
  required int64 port = 3;
}

message HandshakeResponse{
  required bool ok = 1;
  optional int64 givenId = 2;
}

message Module{
  required string type = 1;
  required int64 id = 2;
}

message GetModulesResponse{
  repeated broker.Module module = 1;
}

message SubscribeRequest{
  required int64 idSubscriber = 1;
  repeated int64 id = 2;
}

message SubscribeResponse{
  required bool ok = 1;
  repeated int64 wrongId = 2;
}

message MessageRequest{
  required int64 idRequester = 1;
  optional int64 idReceiver = 2;
  required string message = 3;
}

service BrokerService{
  rpc handshake(broker.HandshakeRequest) returns (broker.HandshakeResponse);
  rpc getModules(broker.EmptyIdRequest) returns (broker.GetModulesResponse);
  rpc subscribe(broker.SubscribeRequest) returns (broker.SubscribeResponse);
  rpc sendMessage(broker.MessageRequest) returns (broker.EmptyMessage);
  rpc close(broker.EmptyIdRequest) returns (broker.EmptyMessage);
}
